### PR TITLE
Add support for TLS SMTP connections.

### PR DESCRIPTION
**Pull request by Logic-32**
I don't want to install an SMTP server on my host so I'd rather use this file for sending emails.  Since I use Amazon's SES, I need TLS support.  There was a TODO to add support so here is the support :)

Before submitting your PR, please review the following checklist:

- [x] **CHOOSE CORRECT BRANCH**: if filing a bugfix/enhancement
      against certain release version, choose `0.9`, `0.10` or `0.11` branch,
      for dev-edition use `master` branch
- [x] **CONSIDER adding a unit test** if your PR resolves an issue
- [x] **LIST ISSUES** this PR resolves
- [x] **MAKE SURE** this PR doesn't break existing tests
- [x] **KEEP PR small** so it could be easily reviewed.
- [x] **AVOID** making unnecessary stylistic changes in unrelated code
- [x] **ACCOMPANY** each new `failregex` for filter `X` with sample log lines
      within `fail2ban/tests/files/logs/X` file

### CODE REVIEW

**Comment by Logic-32** [comment]
Just going to mention #1930 as this will help anyone else looking to use smtp.py with ssl.

File: config/action.d/smtp.py [hunk 1]
```diff
@@ -155,10 +157,19 @@ def _sendMessage(self, subject, text):
...
 		try:
 			self._logSys.debug("Connected to SMTP '%s', response: %i: %s",
 				self.host, *smtp.connect(self.host))
+
+			if self.ssl: # pragma: no cover
+				tls_result = smtp.starttls()[0];
+				if tls_result != 220: # pragma: no cover
+					self._logSys.error(
+						"Failed to starttls() on '%s' for user '%s': %s",
+						self.host, self.user, tls_result)
```
Comment (sebres) [inline config/action.d/smtp.py:170]: Since TLS takes place before login, user is probably irrelevant here, isn't it?

Comment (Logic-32) [reply config/action.d/smtp.py:170]: Definitely a fair point.  I figure the extra information doesn't hurt.  Though, I suppose if someone tries to setup two SMTP connections to the same host, even with different users, both TLS attempts should fail so the user probably doesn't add anything useful.

Want me to remove it?

Comment (sebres) [reply config/action.d/smtp.py:170]: I don't think even the logging is necessarily at all, because raising of error should be enough too.

Comment (Logic-32) [reply config/action.d/smtp.py]: Done.

**Comment by Logic-32** [comment]
@sebres, I'm not sure why some of the checks on this failed.  Is there anything else you need from me to get this merged?

**Comment by sebres** [comment]
Hmm... something wrong going on here with python 2.x (both 2.7 and pypy2 hanging on some test in GHA CI).

**Comment by Logic-32** [comment]
Indeed.  If I'm reading things right, it was `testBan (fail2ban.tests.action_d.test_smtp.SMTPActionTest)` that had the issue.

I'm new to Python, specifically.  If you can give me some guidance on how to run the tests locally and how to account for different python versions I can try to debug/resolve.

**Comment by sebres** [comment]
You can do it directly from the source code root (unbuilt, uninstalled):
```bash
PYTHONPATH=. python2.7 bin/fail2ban-testcases -vv -l 10 'SMTPActionTest'
```
That will run all SMTPActionTests within the test-suite started with python 2.7 with verbosity 2 and heavydebug (10) log-level.

**Comment by sebres** [comment]
The error is that `self.host` cannot be supplied to `SMTP` constructor, because it can contain port... 
but `connect` takes place few lines below, so I guess 57fb7dec8fd772d9900557dfd1f6b1c0f55767fa would fix that.
Just untested with ssl=True (I'm unsure starttls really requires that SMTP instance should become the host by init).

**Comment by Logic-32** [comment]
Thanks for finding and fixing that!  Do you want me to add a test for ssl=true?

**Comment by sebres** [comment]
I'd never say "no" for more code coverage... So yes, if you can.

**Comment by Logic-32** [comment]
Kk.  My personal life has been a bit crazy lately but I hope to take a stab at this in a couple weeks.

**Comment by Logic-32** [comment]
It turns out that [smtpd is old, deprecated](https://docs.python.org/3/library/smtpd.html), and doesn't support `starttls()`.  The [smtpd-tls lib](https://pypi.org/project/smtpd-tls/), however, does allegedly support `starttls()` and is compatible with python 2.7.  But, [aiosmtpd](https://aiosmtpd.readthedocs.io/en/latest/) is the recommended upgrade path.

What is the recommended way to move forward here?  Not add a TLS test or install/migrate to smtpd-tls (since I'm assuming 2.7 support is required)?

**Comment by sebres** [comment]
> I'm assuming 2.7 support is required

It does not. You can skip the TLS test for 2.7 (and a fallback to smtpd_tls as also OK to me).

Otherwise we can also block this unless we remove support for 2.7.

**Comment by Logic-32** [comment]
I now find myself in a situation that shows off just how little I know of how to work with python properly.
- smtpd-tls supports Python 2.7+ but [you can't install it with pip](https://github.com/tintinweb/python-smtpd-tls/issues/1) in part because I hear pip is no longer available on 2.7?  At least, when I just installed it, it installed for python3 only.  I could download/install from source but where would I document that requirement or any other instructions for getting this working?
- aiosmtpd requires Python 3.6+ and f2b supports 3.2.  Granted, this would just be used in tests only so the only problem here is that using aiosmtpd, you wouldn't be able to test with 3.2 either.

Long story short, I'm happy to continue trying to add the tests.  In part because the test itself seems easy.  But, I could use some guidance how how to best resolve the dependency issue with smtpd-tls at least?

How long is it until you plan to remove support for 2.7? :)

**Comment by sebres** [comment]
> aiosmtpd requires Python 3.6+ and f2b supports 3.2

I have no issue with 3.6+ at all. Python 3.6 is released 2016 and is even end-of-life now.
The support of 2.7 is there to retain compatibility for 2.x (which is hardly incompatible to 3.x in many ways), but only to provide possibility to use fail2ban modules with 2.x based software... 
There is simply no 2.8 update for 2.x, but for 3.x there are (compatible) updates.

> I could use some guidance how how to best resolve the dependency issue

Just put the tests in a conditional block:
```python
try:
  import aiosmtpd
except:
  pass
if aiosmtpd: # sys.version_info >= (3,6):
  ...
```

> How long is it until you plan to remove support for 2.7? :)

Well, I don't know... as long as possible, but it is not so far from out.
What doesn't mean that we need implement some enhancements to support that old version.

**Comment by sebres** [comment]
Rebased to master (resolved conflicts in changelog) + few optimizations and code deduplication:
- combine smtpd and aiosmtpd tests;
- encapsulate smtp facilities to setUpClass/tearDownClass (behaves like a singleton, doesn't start smtp server per test);
- don't generate cert every time (too slow by RSA:2048, use pre-generated short ECC:256 instead);

This also simplifying and drastically speeding-up smtp-action tests:
```diff
  $ bin/fail2ban-testcases --fast -vvv SMTPActionTest
  ...
- Ran 7 tests in 4.058s
+ Ran 10 tests in 0.735s
```

**Review by sebres — APPROVED** [review]
looks nice after all, will merge it now
